Introduce state computer protocol updates

LukasGasior1 opened the pull request:
To be added in a follow up PR:
- protocol update execution
- tests for protocol updates \w state version bounds

jakrawcz-rdx commented on core-rust/state-manager/src/protocol/protocol_config.rs:
```diff
@@ -0,0 +1,161 @@
+        support_type: ProtocolUpdateSupportType,
+    },
+    EnactUnconditionallyAtEpoch(Epoch),
+    EnactUnconditionallyAtStateVersion(StateVersion),
```
just as discussed on our call: it still seems to me like some portion of conceptual complexity (and code) could be gone if we discarded the "unconditional" methods (and used `required_ratio_of_stake_supported: 0.0` for "asap updates"; and stored the "already-enacted-by-quorum" updates in a map). Please see if there are any true blockers to migrate to that approach.

jakrawcz-rdx commented on core-rust-bridge/src/main/java/com/radixdlt/protocol/ProtocolUpdateEnactmentBound.java:
```diff
@@ -0,0 +1,82 @@
+
+  record Epoch(UInt64 epoch) implements ProtocolUpdateEnactmentBound {}
+
+  record StateVersion(UInt64 stateVersion) implements ProtocolUpdateEnactmentBound {}
```
I am also a fan of leaving the state versions out.

Possibly there is a way of avoiding them completely (i.e. only ever switch protocol on epoch boundaries) by introducing a way of doing `Engine.forceEpochChangeRightNow()`...?

LukasGasior1 replied:
I removed state version updates for now in a follow up PR: https://github.com/radixdlt/babylon-node/pull/790

jakrawcz-rdx commented on core-rust/state-manager/src/store/rocks_db.rs:
```diff
@@ -206,6 +207,16 @@ impl VersionedCf<Epoch, LedgerProof> for EpochLedgerProofsCf {
 
+/// Ledger proofs of protocol updates.
+/// Schema: `Epoch.to_bytes()` -> `scrypto_encode(VersionedLedgerProof)`
+/// Note: This duplicates a small subset of [`StateVersionToLedgerProof`]'s values.
```
```suggestion
/// Note: This duplicates a small subset of [`LedgerProofsCf`]'s values.
```

_(plz also propagate the fix to my line that you copied :p)_

jakrawcz-rdx commented on core-rust/state-manager/src/store/rocks_db.rs:
```diff
@@ -206,6 +207,16 @@ impl VersionedCf<Epoch, LedgerProof> for EpochLedgerProofsCf {
 }
 
+/// Ledger proofs of protocol updates.
+/// Schema: `Epoch.to_bytes()` -> `scrypto_encode(VersionedLedgerProof)`
```
```suggestion
/// Schema: `StateVersion.to_be_bytes()` -> `scrypto_encode(VersionedLedgerProof)`
```

LukasGasior1 commented:
@talekhinezh @jakrawcz-rdx I'm going to close this PR in favor of https://github.com/radixdlt/babylon-node/pull/790 , which includes the whole feature.
@jakrawcz-rdx I've addressed your comment-related comments there. Let's also discuss unconditional vs 0.0 required support updates there.
